celery_unique/core.py

```diff
@@ -0,0 +1,199 @@
+            the task when it executes @type callback_args: list | tuple
+        :param callback_kwargs: The keyword arguments which will be passed to
+            the task when it executes @type callback_kwargs: dict
+
+        :return: The key which will be used to find any extant version of this
+        task which, if found, will by revoked.  Keys are built by using three
+        colon-delimited components::
+
+            1. A global prefix used to identify that the key/value pair in
+                the backend was created to track
+                a unique Celery task (by default, this is "celery_unique")
+            2. The name of the task (usually the Python dot-notation path to
+                the function)
+            3. The value produced by the `key_generator` callable when supplied
+                with the task's callback arguments.
+
+        :rtype: unicode
+        """
+
+        # Get the unbound lambda used to create `self.unique_key` if the inner
+        # function exists
+        if hasattr(self.unique_key, '__func__'):
+            key_generator = self.unique_key.__func__
+        else:
+            key_generator = self.unique_key
+
+        # Create and return the redis key with the generated unique key suffix
+        return '{prefix}:{task_name}:{unique_key}'.format(
+            prefix=UNIQUE_KEY_PREFIX,
+            task_name=self.name,
+            unique_key=key_generator(*callback_args, **callback_kwargs),
+        )
+
+    @staticmethod
+    def _make_ttl_for_unique_task_record(task_options):
+        """
+        Calculate an aproximate TTL for an enqueued task.
+
+        Given the options provided to `apply_async()` as keyword arguments,
+        determines the appropriate TTL to ensure that a unique task record
+        expires (approximately) at the same time as the earliest time that the
+        task is expected to be executed by Celery.
+
+        The TTL value will be determined by examining the following values, in
+        order of preference:
+            - The `eta` keyword argument passed to `apply_async()`, if any.  If
+              this value is found, then the TTL will be the number of seconds
+              between now and the ETA datetime.
+            - The `countdown` keyword argument passed to `apply_async()`, which
+              will theoretically always exist if `eta` was not provided.  If
+              this value is used, the TTL will be equal.
+
+        Additionally, if an `expires` keyword argument was passed, and its
+        value represents (either as an integer or timedelta) a shorter duration
+        of time than the values provided by `eta` or `countdown`, the TTL will
+        be reduced to the value of `countdown`.
+
+        Finally, the TTL value returned by this method will always be greater
+        than or equal to 1, in order to ensure compatibility with different
+        backend's TTL requirements, and that a record produced for a
```

Comment:
This is super nitpicky, but the apostrophe should come after the `s` in this case:
> ensure compatibility with different backends' TTL requirements

Or we could make it a bit wordier:
> ensure compatibility with TTL requirements of various backends